### DISCUSSION TITLE
VSCode Remote SSH (+ Jupyter Notebook) & Copilot: Is there a way?

### DISCUSSION THREAD

**Discussion by JulesGM** [community/community#6942] [Copilot Conversations] [answered]
I develop almost strictly with VSCode Remote SSH and often in Jupyter Notebook mode (with Python, on a remote GPU cluster). 

Is there a way to use Copilot in this configuration? Thanks.

**Comment by Amocy-Wang** [comment, marked as answer]
When you connect to a ssh remote in vscode, then you can open the Extensions panel. You can find that the Copilot plugin can be chosen to be installed in SSH.

**Reply by flavio87** [reply]
I was able to install github copilot remotely, but then I never got asked to authorize my account and it doesn't seem to work. How to solve this?

**Reply by charmby** [reply]
I came across the same problem

**Reply by ildoonet** [reply]
Me too

**Reply by TechDufus** [reply]
^ Bump

**Reply by PwuJdu** [reply]
> I was able to install github copilot remotely, but then I never got asked to authorize my account and it doesn't seem to work. How to solve this?

You should install copilot in your local environment first.Then login your github account. After that you can connect to the remote ssh and install copilot again then your github account data will be synced.

**Reply by gouchaoer** [reply]
me too

**Reply by Jaieu** [reply]
I installed and checked the settings.json file with "ui" in remote field, but still get an error "manhandled status from server: 500 please login". Anyone knows the answer?

**Comment by thudaniel2333** [comment]
I'm encountering a similar issue. My copilot extension works perfectly locally. But the remote server I'm using is not connected to internet and can only be accessed via ssh. The copilot extension failed to activate because it has to use the connection of the remote server. Is there anyway to solve this issue? e.g. is it possible to make the copilot remote extension use local connection instead?

**Comment by njbrake** [comment]
@thudaniel2333  add this to your settings.json in VSCode and that'll fix it for you:

```
    "remote.extensionKind": {
        "GitHub.copilot": ["ui"],
      }
```

Here's why it works: https://code.visualstudio.com/api/advanced-topics/extension-host

**Reply by aodenweller** [reply]
Thanks a lot. This works for me.

**Reply by dchenhe** [reply]
Thanks a lot. Works for me as well

**Reply by htaodtl** [reply]
Thanks a lot, it works for me

**Reply by Noonlord** [reply]
Thanks a lot!

**Reply by luisrock** [reply]
Great, it works! Thanks!

**Reply by piwawa** [reply]
Awesome, it works like no copilot ever installed, just don't give any single word during coding.

**Reply by JamyDon** [reply]
Thanks a lot! Works pretty well!

**Reply by arkhodamoradi** [reply]
If you are a mac user, you may want to add the following:

"remote.extensionKind": {
        "GitHub.copilot": ["ui"],
        "linhmtran168.mac-ca-vscode": ["ui"],
    },

**Reply by czhcooper** [reply]
Thanks a lot! It works for me, and I'm using MacOS

**Reply by AuroraGazer** [reply]
No, not that case. I have tried this. Making copilot work at local PC does make it able to login. HOWEVER, when I try to use the agent mode, copilot will create a mirror directory locally and do work in it. It seems like that copilot in this case can only edit local files... Do you know any other solutions?

**Reply by M-Mozafary** [reply]
It worked for me as well. I am on MacOS. 
Thanks a lot.

**Comment by huangchieh** [comment]
Just install copilot again after login into remote server works for me.

**Comment by njbrake** [comment]
Update to this thread, if you want github copilot chat to work, you'll have to update it to the following, to add in the copilot-chat feature too

```    
"remote.extensionKind": {
        "GitHub.copilot": ["ui"],
        "GitHub.copilot-chat": ["ui"],
    },
```

**Reply by mrkeaton1** [reply]
This is great! Is there a way to identify the extension name like this? I would like to have this working for git as well.

**Reply by Limezero** [reply]
This was the solution. I'm not sure if it's a recent bug or something that has been broken for years, but Copilot doesn't seem to function at all in a remote environment. `Open User Settings (JSON)` and forcing both extensions to run locally by placing that section at the end fixes it with seemingly no drawbacks.

Thanks to this, I found out that half the stuff I had *also* wasn't working in SSH windows. Due to some utterly bizarre reason, dozens of simple scripts like "take a screenshot of your code" or "increment the currently highlighted number" were disabled, because they "prefer to run in the Remote Extension Host" and demand to be installed there. Debuggers and language servers I can understand, but defaulting to this behavior for everything is ridiculous.

---

Also to answer the previous question, `Extensions panel > Right click on an extension > Copy Extension ID` will get you the IDs required for this. You can also use those to access the Marketplace listing of said extension, or install it by pasting that string into the search bar (very handy to disambiguate which one of the 200 generically named "CSS Snippets" or "JSON" extensions you meant when talking to someone).

**Comment by jamesbooker** [comment]
Question: The settings updates suggested above, should I put them in my local settings.json or the settings.json for the remote host?

Context: When I connect to a remote SSH instance, the copilot chat icon disappears, even though I've installed locally, signed in, then installed in remote. The settings above didn't seem to work for me no matter where I added them

EDIT: Chat disabled by my organisation, causes the icon in the toolbar to disappear
